### PR TITLE
Remove blob

### PR DESCRIPTION
**Pull request by LordAro**
## Motivation / Problem
Unfixable GCC warnings when linking with LTO due to "optimised" code mixing heap & stack objects (it can't ever do this, but GCC warns all the same)
`attempt to free a non-heap object 'hdrEmpty' [-Wfree-nonheap-object]` 

## Description
Blob/CBlobT/CStrA is some weird code (mostly used by YAPF desync dump output) with auto-lifetime managed container objects.
Luckily, the STL exists and we can use std::string/std::vector for the same purposes.

## Limitations
This is basically a 1-1 translation and the code is still real ugly.
Converting "advanced" printf commands (`0x%04X`) to iostream-based code is real ugly. `std::format` would make it nicer, but that requires C++20 :)

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/misc/dbg_helpers.h [hunk 1]
```diff
@@ -67,28 +66,32 @@ inline typename ArrayT<T>::item_t ItemAtT(E idx, const T &t, typename ArrayT<T>:
...
 		for (size_t i = 0; i < ArrayT<T>::length; i++) {
 			if ((value & (1 << i)) == 0) continue;
-			out.AddFormat("%s%s", (out.Size() > 0 ? "+" : ""), (const char*)t[i]);
+			out += (!out.empty() ? "+" : "");
+			out += t[i];
 			value &= ~(E)(1 << i);
 		}
-		if (value != 0) out.AddFormat("%s%s", (out.Size() > 0 ? "+" : ""), t_unk);
+		if (value != 0) {
+			out += (!out.empty() > 0 ? "+" : "");
```
Comment (glx22) [inline src/misc/dbg_helpers.h]: Funny how MSVC annotates each time the function is used, but the warning is very valid `!out.empty() > 0` definitely looks wrong.

File: src/pathfinder/yapf/yapf_costcache.hpp [hunk 2]
```diff
@@ -123,27 +122,26 @@ struct CSegmentCostCacheT : public CSegmentCostCacheBase {
...
+		m_heap.clear();
 	}
 
 	inline Tsegment& Get(Key &key, bool *found)
 	{
 		Tsegment *item = m_map.Find(key);
 		if (item == nullptr) {
 			*found = false;
-			item = new (m_heap.Append()) Tsegment(key);
+			item = &m_heap.emplace_back(key);
```
Comment (frosch123) [inline src/pathfinder/yapf/yapf_costcache.hpp]: It looks to me like SmallArray actually has stable iterators, so the pointer-to-member do not change when growing.
vector does not have that property.
I think we need std::deque for that.

File: src/pathfinder/yapf/nodelist.hpp [hunk 3]
```diff
@@ -24,13 +23,12 @@ class CNodeList_HashTableT {
...
 	typedef Titem_ Titem;                                        ///< Make #Titem_ visible from outside of class.
 	typedef typename Titem_::Key Key;                            ///< Make Titem_::Key a property of this class.
-	typedef SmallArray<Titem_, 65536, 256> CItemArray;           ///< Type that we will use as item container.
 	typedef CHashTableT<Titem_, Thash_bits_open_  > COpenList;   ///< How pointers to open nodes will be stored.
 	typedef CHashTableT<Titem_, Thash_bits_closed_> CClosedList; ///< How pointers to closed nodes will be stored.
 	typedef CBinaryHeapT<Titem_> CPriorityQueue;                 ///< How the priority queue will be managed.
 
 protected:
-	CItemArray      m_arr;        ///< Here we store full item data (Titem_).
+	std::vector<Titem_> m_arr;    ///< Here we store full item data (Titem_).
```
Comment (frosch123) [inline src/pathfinder/yapf/nodelist.hpp]: Probably the same here. std::deque has stable pointer-to-item. std::vector does not.

File: src/pathfinder/yapf/yapf_costcache.hpp (hunk 2 above)
Comment (michicc) [reply src/pathfinder/yapf/yapf_costcache.hpp]: I haven't looked at the code, but just as a note for deque: `In addition, insertion and deletion at either end of a deque never invalidates pointers or references to the rest of the elements.`

Inserting/removing something in the middle will still invalidate pointers.

File: src/misc/dbg_helpers.cpp [hunk 4]
```diff
@@ -56,20 +54,19 @@ static const char * const signal_type_names[] = {
...
 
 
 /** Translate TileIndex into string. */
-CStrA TileStr(TileIndex tile)
+std::string TileStr(TileIndex tile)
 {
-	CStrA out;
-	out.Format("0x%04X (%d, %d)", tile, TileX(tile), TileY(tile));
-	return out.Transfer();
+	std::stringstream ss;
```
Comment (frosch123) [inline src/misc/dbg_helpers.cpp:65]: personally I always do "char buf[32]" and "seprintf" in such cases.
But now you already put in the effort to find those weird stream modifiers :p

**Review by frosch123 — APPROVED** [review]

File: src/misc/dbg_helpers.cpp (hunk 4 above)
Comment (LordAro) [reply src/misc/dbg_helpers.cpp:65]: Depends how much we care about the extra overhead of sstream/iomanip includes

**Pull request merged**